csrc/multidevice/communication.cpp

```diff
@@ -40,56 +73,125 @@ inline void assertBuffersHaveSameSize(
   }
 }
 
-inline void post_common(Communication& self, Communicator& comm) {
-  NVF_ERROR(
-      std::find(
-          self.params().team.begin(),
-          self.params().team.end(),
-          comm.deviceId()) != self.params().team.end(),
-      "current device index ",
-      comm.deviceId(),
-      " must be present in the communication's team");
-}
-
 inline void doLocalCopy(const at::Tensor& dst, const at::Tensor& src) {
   dst.copy_(src, /* non-blocking */ true);
 }
 
-} // namespace
+inline bool hasRoot(CommunicationType type) {
+  return type == CommunicationType::Gather ||
+      type == CommunicationType::Scatter ||
+      type == CommunicationType::Broadcast ||
+      type == CommunicationType::SendRecv;
+}
+
+inline bool isReduction(CommunicationType type) {
+  return type == CommunicationType::Reduce ||
+      type == CommunicationType::Allreduce ||
+      type == CommunicationType::ReduceScatter;
+}
 
-Communication::Communication(CommParams params, std::string name, bool has_root)
-    : params_(std::move(params)),
-      collective_type_(std::move(name)),
-      has_root_(has_root) {
-  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
+inline void assertValid(
+    const CommParams& params,
+    const DeviceIdxType my_device_index) {
+  assertBuffersHaveSameSize(params.src_bufs, params.dst_bufs);
+  std::unordered_set<DeviceIdxType> team_without_duplicates(
```

Comment:
Is DeviceIdxType the device id or the index of a device in the Mesh/Team? I know it's not caused by this PR, but I think DeviceIdxType is used as both and can cause confusion.
For example, the later function `DeviceIdxType getRootRelativeIndex` uses a DeviceIdxType root to calculate the relative index which is also a DeviceIdxType. 

wdyt about something like a `DeviceId` and `DeviceIndex`?

DeviceId root = 3;
std::vector<DeviceId> team = {1, 3, 10};
DeviceIndex root_index = getRootRelativeIndex(...) // 1

Reply:
Also, since this isn't directly related to the PR, we can address this later

Reply:
> Is DeviceIdxType the device id or the index of a device in the Mesh/Team?

That would definitely be confusing. Can you remind me when we use it to mean the latter? I only saw it used to mean the former.

Reply:
Very good point. I fixed that by changing the return type of `getRootRelativeIndex`. I think this is the only relevant place

> Can you remind me when we use it to mean the latter? I only saw it used to mean the former.

the "root" passed to process group needs to be the relative index of the deviceId within the Team. Does it make sense ?